### PR TITLE
Create TestCaseMixin for Helper Functions Across All Test Case Bases

### PR DESCRIPTION
**Pull request by smk4664**
…cess to the same helper functions.

# Closes: #1949 

Adds TestCaseMixin and moves the functions from TestCase to the Mixin.

# TODO

- [x] Explanation of Change(s)
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/utilities/testing/__init__.py [hunk 1]
```diff
@@ -114,6 +117,16 @@ def setUp(self):
...
+        # Create the test user and assign permissions
+        self.user = User.objects.create_user(username="testuser")
+        self.add_permissions(*self.user_permissions)
+
+        # Initialize the test client
+        self.client = Client()
+
+        # Force login explicitly with the first-available backend
+        self.client.force_login(self.user)
+
```
Comment (glennmatthews) [inline nautobot/utilities/testing/__init__.py:129]: Any reason we couldn't add a `setUp()` to TestCaseMixin that centralizes this logic as well?

Comment (smk4664) [reply nautobot/utilities/testing/__init__.py:129]: Probably. When I was working with Jathan we kept it out, but I will try with the change.

File: nautobot/utilities/testing/mixins.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+from taggit.managers import TaggableManager
+
+from nautobot.extras.models import Tag
+from nautobot.users.models import ObjectPermission
+from nautobot.utilities.permissions import resolve_permission_ct
+from nautobot.utilities.fields import JSONArrayField
+from .utils import extract_form_failures
+
+
+class TestCaseMixin:
```
[the remark this answers is not in the document]
Comment (smk4664) [reply nautobot/utilities/testing/mixins.py:18]: I can look into integrating with Integration test case, I haven't spent much time with that one. I think having the word `Nautobot` would be a little redundant, but it wouldn't hurt.

**Comment by bryanculver** [comment]
@smk4664 Can we have this target `next` instead of `develop`? Hopefully minimize the potential impact to developers and CI on patch releases for those doing extensive jobs testing.

**Comment by smk4664** [comment]
> @smk4664 Can we have this target `next` instead of `develop`? Hopefully minimize the potential impact to developers and CI on patch releases for those doing extensive jobs testing.

I guess, but I really need this in order to fix tests for Nautobot-plugin-welcome-wizard.  We use TransactionTestCases in one of the views because it calls a job, and the TransactionTestCase does not have access to the helper functions I am trying to add here in the mixin.  This worked when testing prior to Nautobot 1.3.x as Nautobot 1.3.0 requires that Job tests use TransactionTestCase.  That is why I categorized the ticket as a bug and I am trying to introduce the fix in 1.3.x instead of having to wait for 1.4.

**Comment by bryanculver** [comment]
> > @smk4664 Can we have this target `next` instead of `develop`? Hopefully minimize the potential impact to developers and CI on patch releases for those doing extensive jobs testing.
> 
> I guess, but I really need this in order to fix tests for Nautobot-plugin-welcome-wizard. We use TransactionTestCases in one of the views because it calls a job, and the TransactionTestCase does not have access to the helper functions I am trying to add here in the mixin. This worked when testing prior to Nautobot 1.3.x as Nautobot 1.3.0 requires that Job tests use TransactionTestCase. That is why I categorized the ticket as a bug and I am trying to introduce the fix in 1.3.x instead of having to wait for 1.4.

The concern is that since this is missing helper functions (`TransactionTestCase` itself isn't broken, it just lacks implementation of all methods of `TestCase`) that in the meantime other users implemented this in setup themselves and that may pose conflicts. Their CI may fail on a patch release.

File: nautobot/utilities/testing/mixins.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+    user_permissions = ()
+
+    def setUpNautobot(self, client=True):
+        """Setup shared testuser, statuses and client."""
+        # Re-populate status choices after database truncation by TransactionTestCase
+        populate_status_choices(apps, None)
+
+        # Create the test user and assign permissions
+        self.user = User.objects.create_user(username="testuser")
```
Comment (glennmatthews) [inline nautobot/utilities/testing/mixins.py:37]: Just to minimize the risk of causing collisions with plugins that might have copied-and-pasted the original code, let's change this username to something like `"nautobotuser"` as part of this changeset.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/utilities/testing/mixins.py (hunk 3 above)
Comment (smk4664) [reply nautobot/utilities/testing/mixins.py]: Done!

**Pull request merged**